Test TLS 1.2 builds with each encryption type

mprse opened the pull request:
## Description
Resolves https://github.com/Mbed-TLS/mbedtls/issues/6313


## Status
**READY**

## Requires Backporting
Yes/2.28
https://github.com/Mbed-TLS/mbedtls/pull/6394

## Migrations
NO

## Additional comments
Any additional information that could be of interest

## Todos
- [X] null cipher
- [X] only CBC-legacy (no EtM)
- [X] only CBC-legacy and CBC-EtM

gilles-peskine-arm commented:
> I assume that the check fails correctly as AEAD is disabled, so probably this test should be skipped.

Right, it makes sense that `mbedtls_ssl_context_save` returns a “I can't do this at all” error in this situation, which has precedence over the usual “I could do this if the buffer was bigger”. Since handshake serialization is only implemented with AEAD, there should probably be a requirement in `check_config.h` (and then `MBEDTLS_SSL_CONTEXT_SERIALIZATION` should be disabled in this non-AEAD build).

> is MBEDTLS_SSL_SOME_SUITES_USE_MAC guard correct

Looks correct to me: all supported cipher suites either have a MAC or an AEAD tag (we don't implement the NULL_NULL cipher suite which has no security at all, only cipher suites with authentication but where the encryption may be NULL). If there are MAC cipher suites in the build then we check whether the selected cipher suite uses AEAD; if there are no MAC cipher suites then the selected cipher suite must be AEAD.

mpg commented on library/ssl_ticket.c:
```diff
@@ -114,6 +114,7 @@ static int ssl_ticket_gen_key( mbedtls_ssl_ticket_context *ctx,
 /*
  * Rotate/generate keys if necessary
  */
+#if defined(MBEDTLS_CIPHER_MODE_AEAD) || defined(MBEDTLS_NIST_KW_C)
```
I understand this makes sense as this is what guards `mbedtls_cipher_auth_encrypt_ext()` but in this context only AEAD is allowed - `mbedtls_ssl_ticket_setup()` enforces that.

mprse replied:
After optimizing the configurations I think this can be reverted.

mpg commented on library/psa_crypto.c:
```diff
@@ -3592,6 +3592,7 @@ static psa_status_t psa_aead_check_nonce_length( psa_algorithm_t alg,
             break;
 #endif /* PSA_WANT_ALG_CHACHA20_POLY1305 */
         default:
+            (void) nonce_length;
```
Should we guard this cast with `#if !GCM && !CCM && !ChachaPoly` or is it overkill?

mprse replied:
Hmm it doesn't affect the binary size (just silent compiler), so it looks ok for me.

mpg commented on tests/scripts/all.sh:
```diff
@@ -1277,6 +1277,7 @@ component_test_crypto_full_no_cipher () {
 component_test_crypto_default_stream_cipher_only () {
     msg "build: default with only stream cipher"
 
+    scripts/config.py crypto_full
```
I think it should be `full` not `crypto_full`. The motivation for this is really to catch issues in TLS 1.2 code.

mprse replied:
I noticed that in full config tls 1.2 is disabled so it makes no sense to test it here instead I added psa and no psa configs.

mpg replied:
I think that's because you're using `crypto_full` not `full`. (I just tried, `full` leave TLS 1.2 enabled for me.) As the name implies, `crypto_full` only enables crypto (no X.509, no TLS).

That said, I'm happy with testing "default" and "default + use_psa" as well. But `crypto_full` was wrong.

mprse replied:
Yes, that is the case. I provided `crypto_full`  by mistake (copied wrong line) and treated as full. But if `default` with `psa/no psa` is sufficient I won't add full here.

mpg reviewed (changes requested):
Just a first quick review for early feedback.

mprse commented:
While addressing review comments and optimizing components setup I encountered strange failure:
```
CMAC init #7 Camellia: wrong cipher ............................... FAILED
  ( cipher_info = mbedtls_cipher_info_from_type( cipher_type ) ) != NULL
  at line 112, suites/test_suite_cmac.function

```

Test case:
https://github.com/Mbed-TLS/mbedtls/blob/77c691f099f820a3edc1f330c3c10936dda88c18/tests/suites/test_suite_cmac.data#L31-L33

`MBEDTLS_CIPHER_ID_CAMELLIA` is cipher id and it is used as cipher type? This is wrong usage. It is probably done to achieve `MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA` returned by `mbedtls_cipher_cmac()`, but now `mbedtls_cipher_info_from_type()` fails.
`MBEDTLS_CIPHER_ID_CAMELLIA` is treated as `MBEDTLS_CIPHER_AES_128_CBC` which in stream cipher only config is disabled.

The test case description says `CMAC init #7 Camellia: wrong cipher `, but it is unclear to me that we are testing here.

mpg commented:
Ah, each time we add tests we uncover new and in this case pretty unexpected bugs!

I think test cases 5 to 7 are wrong in that they use a cipher ID where a cipher type is expected, as you say. I think cases 5 and 6 should use `MBEDTLS_CIPHER_AES_128_ECB` and case 7 should use `MBEDTLS_CIPHER_CAMELLIA_128_ECB` - this test is about checking that only AES and 3DES are accepted (because CMAC is defined by NIST, and AES and 3DES are the only NIST-approved block ciphers).

mprse commented on tests/scripts/all.sh:
```diff
@@ -1274,6 +1274,38 @@ component_test_crypto_full_no_cipher () {
+    make
+
+    msg "test: default with only stream cipher"
+    make test
```
I tried to run `ssl-opt.sh` with `TLS 1.2` filter to limit executed tests and it works for `CBC` and `CBC-EtM`, but for `Stream/NULL` not and I wonder if this make sense for Stream/NULL:

```
******************************************************************
* test_tls1_2_default_stream_cipher_only: test: default with only stream cipher - ssl-opt.sh (subset) 
* czw, 29 wrz 2022, 13:53:24 CEST
******************************************************************
CBC Record splitting: TLS 1.2, no splitting ............................ SKIP
Certificate hash: client TLS 1.2 -> SHA-2 .............................. PASS
Small client packet TLS 1.2 BlockCipher ................................ SKIP
Small client packet TLS 1.2 BlockCipher, without EtM ................... SKIP
Small client packet TLS 1.2 BlockCipher larger MAC ..................... SKIP
Small client packet TLS 1.2 AEAD ....................................... SKIP
Small client packet TLS 1.2 AEAD shorter tag ........................... SKIP
Small client packet DTLS 1.2 ........................................... SKIP
Small client packet DTLS 1.2, without EtM .............................. SKIP
Small server packet TLS 1.2 BlockCipher ................................ SKIP
Small server packet TLS 1.2 BlockCipher, without EtM ................... SKIP
Small server packet TLS 1.2 BlockCipher larger MAC ..................... SKIP
Small server packet TLS 1.2 AEAD ....................................... SKIP
Small server packet TLS 1.2 AEAD shorter tag ........................... SKIP
Small server packet DTLS 1.2 ........................................... SKIP
Small server packet DTLS 1.2, without EtM .............................. SKIP
Large client packet TLS 1.2 BlockCipher ................................ SKIP
Large client packet TLS 1.2 BlockCipher, without EtM ................... SKIP
Large client packet TLS 1.2 BlockCipher larger MAC ..................... SKIP
Large client packet TLS 1.2 AEAD ....................................... SKIP
Large client packet TLS 1.2 AEAD shorter tag ........................... SKIP
Large server packet TLS 1.2 BlockCipher ................................ SKIP
Large server packet TLS 1.2 BlockCipher, without EtM ................... SKIP
Large server packet TLS 1.2 BlockCipher larger MAC ..................... SKIP
Large server packet TLS 1.2 BlockCipher, without EtM, truncated MAC .... SKIP
Large server packet TLS 1.2 AEAD ....................................... SKIP
Large server packet TLS 1.2 AEAD shorter tag ........................... SKIP
DTLS fragmenting: gnutls server, DTLS 1.2 .............................. RETRY(client-timeout) RETRY(client-timeout) FAIL
  ! bad client exit code (expected 0, got 143)
  ! outputs saved to o-XXX-28.log
DTLS fragmenting: gnutls client, DTLS 1.2 .............................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-29.log
DTLS fragmenting: openssl server, DTLS 1.2 ............................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-30.log
DTLS fragmenting: openssl client, DTLS 1.2 ............................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-31.log
DTLS fragmenting: 3d, gnutls server, DTLS 1.2 .......................... SKIP
DTLS fragmenting: 3d, gnutls client, DTLS 1.2 .......................... SKIP
DTLS fragmenting: 3d, openssl server, DTLS 1.2 ......................... SKIP
DTLS fragmenting: 3d, openssl client, DTLS 1.2 ......................... SKIP
TLS 1.3: Not supported version check:gnutls: srv max TLS 1.2 ........... SKIP
TLS 1.3: Not supported version check:openssl: srv max TLS 1.2 .......... SKIP
TLS 1.2: Check rsa_pss_rsae compatibility issue, m->O .................. SKIP
TLS 1.2: Check rsa_pss_rsae compatibility issue, m->G .................. SKIP
------------------------------------------------------------------------
FAILED (35 / 39 tests (34 skipped))

```

Most are skipped and fails are caused because cipher-suits do not match:
```
Error in handshake(): No supported cipher suites have been found.
Error in handshake(): No supported cipher suites have been found.
Error in handshake(): No supported cipher suites have been found.
Error in handshake(): No supported cipher suites have been found.
Error in handshake(): No supported cipher suites have been found.
Exiting via signal 15
-------
ssl_tls12_server.c:1683: |1| got no ciphersuites in common
-------
ssl_tls12_server.c:1683: |1| got no ciphersuites in common
-------
140021506344256:error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher:../ssl/statem/statem_srvr.c:2283:
```

mpg replied:
Ah, I assume gnutls and openssl don't offer/accept NULL ciphersuites by default (which is of course extremely sensible as these ciphersuite do not provide any confidentiality), so evidently trying to interop with them in a build where we only support NULL ciphersuite is not going to work well.

I'm not sure how to properly express that in terms of `requires_config_enabled` dependencies... Perhaps the simplest thing to do is to just invoke `ssl-opt.sh` with `-e gnutls|openssl` in this component?

mpg replied:
Hmm, I don't understand, the CI seems to be passing now. What changed?

mpg replied:
Oh I see now, `ssl-opt.sh` was added to the other components but not this one. My suggestion stands, then (with a comment explaining why). (Also, perhaps it's `-e 'gnutls\|openssl` actually. And perhaps we need to also exclude things like `m->O`, `G->m` etc.)

mprse replied:
Yes, I removed this because even if we skip `gnutls\|openssl` (4 failures) then only one test will be executed: `Certificate hash: client TLS 1.2 -> SHA-2`.

mpg replied:
Ah, good point. Probably best to leave a comment then so that we don't wonder about it in the future, for example "Not running ssl-opt.sh because most tests require a non-NULL ciphersuite."

gilles-peskine-arm replied:
`ssl-opt.sh` was originally written with a “normal” configuration in mind, using certificates. Until relatively recently, it couldn't do anything useful in a PSK-only build. Then I added instrumentation to adapt the command lines of `CLI_CMD` and `SRV_CMD` to pass `-psk …` or similar so that test cases that aren't specifically related to asymmetric cryptography can run. For null ciphersuites, we'd need similar instrumentation to add some form of `-allow-null-ciphersuites`. I don't know if it's worth the effort.

mpg replied:
Good point, Gilles, but I don't think it's worth the effort for now. NULL ciphersuites are quite niche, so I don't think we need to pro-actively improve our support/testing for them; at least that was not the goal of this issue, but rather to catch an fix mistakes in `#ifdef`s that would have been a distraction when working on builds with driver-only ciphers/AEAD, ie start this work from a clean(er) state.

mpg reviewed (changes requested):
Looks pretty good to me, just one minor point about a runtime dependency that we probably should document and enforce now that we noticed it, even if it doesn't seem to be causing issues here.

mpg commented on include/mbedtls/check_config.h:
```diff
@@ -962,7 +962,9 @@
 #endif
 
-
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION) && !( defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C) || defined(MBEDTLS_CHACHAPOLY_C) )
```
This should also be documented in `mbedtls_config.h`. As a general rule, we need to remember to keep those two files in sync.

mpg commented on tests/scripts/all.sh:
```diff
@@ -1274,6 +1274,159 @@ component_test_crypto_full_no_cipher () {
+}
+
+component_test_tls1_2_default_stream_cipher_only_use_psa () {
+    msg "build: deafult with only stream cipher use psa"
```
Typo: default (here and before make test).

mpg commented on include/mbedtls/check_config.h:
```diff
@@ -962,6 +962,11 @@
 #endif
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) && \
+    !( defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C) || defined(MBEDTLS_CHACHAPOLY_C) || defined(MBEDTLS_NIST_KW_C) )
```
I'm sorry, but I really think `NIST_KW` shouldn't be part of the list, see my [previous comment](https://github.com/Mbed-TLS/mbedtls/pull/6374/files/8d4b24102883315d32135af4e30f5857ee14c582..ce5b68c7a304ad5c9b15613925ae47f0b923ae15#diff-4fc3811c7c53e1746e1056ac38a6462ad2160077a59898bbd2bc31f5938ac185).

mpg commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -1701,6 +1703,9 @@
  * callbacks are provided by MBEDTLS_SSL_TICKET_C.
  *
+ * Requires: MBEDTLS_GCM_C or MBEDTLS_CCM_C or MBEDTLS_CHACHAPOLY_C or
+ *           MBEDTLS_NIST_KW_C
```
Same.

mpg reviewed (changes requested):
Almost there, but I still disagree about one dependency.

mpg reviewed (dismissed):
Thanks for addressing my feedback, looks all good to me now!

mprse commented:
I missed ChangeLog entry and need to add backport as it is approved now.
> The test and fixes need to be backported to 2.28.

It looks to me that entire PR should be backported.

mpg commented:
> It looks to me that entire PR should be backported.

Indeed, looks the same way to me!

AndrzejKurek commented on library/constant_time.c:
```diff
@@ -81,7 +81,7 @@ unsigned mbedtls_ct_uint_mask( unsigned value )
 }
 
-#if defined(MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC)
+#if defined(MBEDTLS_SSL_SOME_SUITES_USE_MAC)
```
It's interesting that we came to [simillar results](https://github.com/Mbed-TLS/mbedtls/pull/6385/files#diff-63a9649f9ecb93f17d935dfdc33a6b129ae5f80e9f675313bf3869eb46e40db8R217), although mine seem to be an overkill :)

AndrzejKurek commented on tests/suites/test_suite_cmac.data:
```diff
@@ -22,15 +22,15 @@ mbedtls_cmac_setkey:MBEDTLS_CIPHER_DES_EDE3_ECB:192:0
 CMAC init #5 AES-224: bad key size
 depends_on:MBEDTLS_AES_C
-mbedtls_cmac_setkey:MBEDTLS_CIPHER_ID_AES:224:MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA
+mbedtls_cmac_setkey:MBEDTLS_CIPHER_AES_128_ECB:224:MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA
```
This will cause a conflict with https://github.com/Mbed-TLS/mbedtls/pull/6385/files#diff-b55f0cc7bb0504cadf4963d58857b281f7af712a7f5587e563a234d11bb4abd3R33, but we can handle it later. Just leaving here to remember :)

AndrzejKurek replied:
OTOH, shouldn't this be MBEDTLS_CIPHER_AES_128_CBC?

gilles-peskine-arm replied:
ECB seems right. The design of CMAC is close to CBC, but it's a slightly tweaked CBC and CMAC isn't implemented on top of CBC, it's a separate implementation on top of what we call ECB but is actually just operating on one block at a time.

AndrzejKurek replied:
Oh, OK, I just found [here ](https://github.com/Mbed-TLS/mbedtls/pull/6385/files#diff-b55f0cc7bb0504cadf4963d58857b281f7af712a7f5587e563a234d11bb4abd3R32) that there test #7 (CMAC init #7 Camellia: wrong cipher) is failing without CBC so I figured it had to be it.

gilles-peskine-arm replied:
There too, CBC is not the right dependency. Neither the test code nor the implementation of CMAC refer to CBC, so I expect it's some other configuration option. `cmac.c` only supports CMAC over DES and AES, not over Camellia or ARIA, and it's not very clean about this, so maybe the problem is with the block cipher and not with the mode?

mprse replied:
I made this change because there was a mismatch between cipher id and cipher type (https://github.com/Mbed-TLS/mbedtls/pull/6374#issuecomment-1261867239). @AndrzejKurek do you expect some changes here?

mpg replied:
Another argument why ECB is correct: it's what we use for the positive cases (1 to 4 just above). And of course the negative cases should be as close as possible to the positive ones.

AndrzejKurek replied:
OK, I can see that it's because `MBEDTLS_CIPHER_ID_CAMELLIA` (`cipher_id_t`) was used instead of a `mbedtls_cipher_type_t` type, `MBEDTLS_CIPHER_CAMELLIA_192_ECB`, and coincidentally, `MBEDTLS_CIPHER_ID_CAMELLIA` translates to 5, the same as `MBEDTLS_CIPHER_AES_128_CBC`.

AndrzejKurek reviewed (commented):
Looks good, but I just had one question about CBC vs ECB in cmac. Not sure if we want to test with the same thing as before, and even if this was the same thing, or it doesn't matter.

mpg commented:
Thanks for reviewing @AndrzejKurek! Can you also adjust the labels? Thanks!

AndrzejKurek commented:
> Thanks for reviewing @AndrzejKurek! Can you also adjust the labels? Thanks!

Thanks for reminding me, I have to put a big sticker on my monitor to remember next time.

mpg reviewed (dismissed):
LGTM

mpg commented on include/mbedtls/check_config.h:
```diff
@@ -962,7 +962,14 @@
 
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) && \
+    !( defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C) || defined(MBEDTLS_CHACHAPOLY_C) )
+#error "MBEDTLS_SSL_SESSION_TICKETS defined, but not all prerequisites"
```
Hmm, sorry, can you remind me why `SSL_SESSION_TICKETS`, as opposed to `SSL_TICKET_C`, has a dependency on AEAD? I'm afraid we might have done the wrong thing here (and I'm afraid it was based on a suggestion from me, sorry if that's the case.)

mprse replied:
From the commit message:
> Both functions are calling mbedtls_cipher_auth_[encrypt/decrypt]_ext() functions. These functions are guarded with MBEDTLS_CIPHER_MODE_AEAD || MBEDTLS_NIST_KW_C flags - make it consistent.
As a result ssl_server2 won't build now with MBEDTLS_SSL_SESSION_TICKETS enabled (mbedtls_cipher_auth_[encrypt/decrypt]_ext() functions not available).
Mark MBEDTLS_SSL_SESSION_TICKETS as dependent on MBEDTLS_CIPHER_MODE_AEAD || MBEDTLS_NIST_KW_C and disable MBEDTLS_SSL_SESSION_TICKETS in stream cipher only build.

Then `MBEDTLS_CIPHER_MODE_AEAD` was switched to `defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C) || defined(MBEDTLS_CHACHAPOLY_C)` as it couldn't be used in `check_config.h`.

Then `MBEDTLS_NIST_KW_C` was removed.

mprse replied:
- `mbedtls_cipher_auth_[encrypt/decrypt]ext()` functions depend on `AEAD (MBEDTLS_GCM_C || MBEDTLS_CCM_C || MBEDTLS_CHACHAPOLY_C)`
- These functions are used in `ssl_msg.c` (guarded with `MBEDTLS_GCM_C || MBEDTLS_CCM_C || MBEDTLS_CHACHAPOLY_C`) and `ssl_ticket.c`.
- As you mentioned [here](https://github.com/Mbed-TLS/mbedtls/pull/6374#discussion_r984300455) `SSL_SESSION_TICKETS` should depend on `AEAD` and `GCM || CCM || ChachaPoly`, but you refered to `SSL_TICKET_C` (not `SSL_SESSION_TICKETS`).
- We can't build with `MBEDTLS_SSL_SESSION_TICKETS` and without `SSL_TICKET_C`. So `MBEDTLS_SSL_SESSION_TICKETS` also depends on `GCM || CCM || ChachaPoly`.

Summary:
For me it should be fixed as follows:
- `SSL_TICKET_C` depends on `GCM || CCM || ChachaPoly`.
- `MBEDTLS_SSL_SESSION_TICKETS` depends on  `SSL_TICKET_C`.

AndrzejKurek replied:
This approach seems right to me.

mprse replied:
Ok but in this case `ssl_server2.c` should use `MBEDTLS_SSL_SESSION_TICKETS && SSL_TICKET_C` guards. Linker undefined references:
```
ssl_server2.c:(.text.startup+0x17b): undefined reference to `mbedtls_ssl_ticket_init'
/usr/bin/ld: ssl_server2.c:(.text.startup+0x93d): undefined reference to `mbedtls_ssl_ticket_free'
/usr/bin/ld: ssl_server2.c:(.text.startup+0x1f1c): undefined reference to `mbedtls_ssl_ticket_setup'
/usr/bin/ld: ssl_server2.c:(.text.startup+0x1f32): undefined reference to `mbedtls_ssl_ticket_parse'
/usr/bin/ld: ssl_server2.c:(.text.startup+0x1f39): undefined reference to `mbedtls_ssl_ticket_write'
/usr/bin/ld: ssl_server2.c:(.text.startup+0x1fcd): undefined reference to `mbedtls_ssl_ticket_rotate'
```

mpg replied:
Yes, indeed I think the current guards are wrong and `MBEDTLS_SSL_SESSION_TICKETS && SSL_TICKET_C` would be correct.

mprse commented:
Added also reworded change log entry proposed by @AndrzejKurek in backport PR. I will continue working on backport when this one is fully accepted.

mpg reviewed (approved):
LGTM!

AndrzejKurek commented:
Unfortunately the CI failed, it seems that session resumption tests that use tickets don't have the proper dependencies.

mprse commented:
> Unfortunately the CI failed, it seems that session resumption tests that use tickets don't have the proper dependencies.

Not `SSL_TICKET_C` but `MBEDTLS_SSL_TICKET_C` in  `fuzz_server.c, ssl_server2.c`. Force-pushed the fix.

mpg reviewed (dismissed):
LGTM

AndrzejKurek commented on include/mbedtls/check_config.h:
```diff
@@ -874,6 +874,11 @@
 
+#if defined(MBEDTLS_SSL_TICKET_C) && \
+    !( defined(MBEDTLS_GCM_C) || defined(MBEDTLS_CCM_C) || defined(MBEDTLS_CHACHAPOLY_C) )
+#error "MBEDTLS_SSL_TICKET_C  defined, but not all prerequisites"
```
Minor: double space

AndrzejKurek commented on programs/ssl/ssl_server2.c:
```diff
@@ -1404,7 +1404,7 @@ int main( int argc, char *argv[] )
     mbedtls_ssl_cache_context cache;
 #endif
-#if defined(MBEDTLS_SSL_SESSION_TICKETS)
+#if defined(MBEDTLS_SSL_SESSION_TICKETS) && defined(MBEDTLS_SSL_TICKET_C)
```
Ifdef at line 286 - missing `MBEDTLS_SSL_TICKET_C`

mpg replied:
@mprse I believe it's about the guards around `USAGE_TICKETS` near the top of the file. It's not a major issue (the worst that will happen is that the usage message will mention options that are not actually available in some configs), but not that Andrzej spotted it, we might as well fix it.

mprse replied:
Yup. I hope it is ok now (fore-pushed last commit).

AndrzejKurek reviewed (changes requested):
Please fix the missing ifdef around the usage text.

mpg reviewed (approved):
LGTM

mprse commented:
This one together with backport seems to be ready for merge, but I'm not sure about the CI. There is windows build failure, but looks unrelated.

gilles-peskine-arm commented:
The Windows failure on OpenCI is an infrastructure glitch and the armcc failure is a known infrastructure issue. The Arm CI passed so this PR has passed the CI.
